src/control/build/interop.go

```diff
@@ -0,0 +1,165 @@
+type VersionedComponent struct {
+	Version   Version
+	Component Component
+}
+
+func (vc *VersionedComponent) String() string {
+	return fmt.Sprintf("%s:%s", vc.Component, vc.Version)
+}
+
+// InteropRule is a rule for checking compatibility between two
+// versioned components.
+type InteropRule struct {
+	Self        Component
+	Other       Component
+	Description string
+	CheckFn     func(self, other *VersionedComponent) bool
+}
+
+// defaultRules are a set of default rules which should apply regardless
+// of release or caller.
+var defaultRules = []InteropRule{
+	// For the 2.0/2.2 releases, at least, the servers must be
+	// be patch-compatible (i.e. the major/minor versions must
+	// be the same). This rule may need to be relaxed or modified
+	// for future releases to allow servers to interoperate
+	// across minor versions.
+	{
+		Self:        ComponentServer,
+		Other:       ComponentServer,
+		Description: "server/server must be patch-compatible",
+		CheckFn: func(self, other *VersionedComponent) bool {
+			return self.Version.PatchCompatible(other.Version)
+		},
+	},
+	{
+		Description: "no backward compatibility prior to 2.0.0",
+		CheckFn: func(self, other *VersionedComponent) bool {
+			v2_0 := MustNewVersion("2.0.0")
+			return !((self.Version.Equals(v2_0) || self.Version.GreaterThan(v2_0)) &&
+				other.Version.LessThan(v2_0))
+		},
+	},
+}
+
+// CheckCompatibility checks a pair of versioned components
+// for compatibility based on specific interoperability constraints
+// or general rules.
+//
+// The design is aimed at allowing a component to verify compatibility
+// with another component, and the logic can be customized by callers
+// for specific requirements at the call site.
+//
+// e.g. "I am server v2.0.0. Am I compatible with agent v1.2.0?"
+func CheckCompatibility(self, other *VersionedComponent, customRules ...InteropRule) error {
+	if self == nil || other == nil {
+		return errors.New("nil components")
+	}
+
+	// If the versions are equal, there's nothing else to check.
+	if self.Version.Equals(other.Version) {
+		return nil
+	}
+
+	// Apply flexible rule-based checks first.
+	for _, rule := range append(customRules, defaultRules...) {
+		if rule.Self.Matches(self.Component) && rule.Other.Matches(other.Component) {
+			if !rule.CheckFn(self, other) {
+				return errors.Wrap(errIncompatComponents(self, other), rule.Description)
+			}
+		}
+	}
+
+	// NB: The following rules loosely follow the semantic versioning spec.
+	// https://semver.org/
+
+	// If only the patch versions are different, then by default
+	// we assume that the components are compatible.
+	if self.Version.PatchCompatible(other.Version) {
+		return nil
+	}
+
+	// If the major versions are beyond the max allowed delta,
+	// the components are incompatible.
+	if self.Version.MajorDelta(other.Version) > MaxMajorDelta {
+		return errors.Wrap(errIncompatComponents(self, other), "major version delta too large")
+	}
```

Comment:
Yeah, I was trying to leave things somewhat flexible to handle non-standard requirements, but maybe it would be best to keep these hard-coded rules as simple as possible.

Another thought that occurred to me as I was re-reviewing this the other day was that I could remove these hard-coded rules altogether and just add them as the last set of default flexible rules.

Reply:
If we need to accommodate multiple major versions at some point, the checks will probably get more complex than this anyway.

I think it would make sense to include them in flexible rules.